### PR TITLE
host-dma: only do DMA L1 exit when data is copied

### PR DESCRIPTION
**Pull request by keyonjie**
Exiting the DMA L1 state, e.g. on a timer tick when there is no data to
copy, prevents the system from staying in S0ix power states. To extend
S0ix residency, this L1 exit needs to happen only after data copied.

Register host DMA usage when there is real data to be copied, for both
timer scheduled and DMA interrupt scheduled cases.

Add a l1_exit_needed field to hda_chan_data, to indicate if L1 exit is
needed at LL scheduler post run. Release the host DMA usage at the post
run notifier callback, to perform the DMA L1 exit when needed.

With these changes applied, the S0ix residency will be extended, and
only performing the DMA L1 exit at LL scheduler post run will preserve
the optimization of the low MCPS usage at the same time.

Signed-off-by: Keyon Jie <yang.jie@linux.intel.com>

### CODE REVIEW

File: src/drivers/intel/hda/hda-dma.c [hunk 1]
```diff
@@ -309,9 +311,14 @@ static void hda_dma_post_copy(struct dma_chan_data *chan, int bytes)
 		 */
 		hda_dma_inc_fp(chan, bytes);
 
-		/* Force Host DMA to exit L1 if scheduled on DMA */
+		/*
+		 * Force Host DMA to exit L1 if scheduled on DMA,
+		 * otherwise, perform L1 exit at LL scheduler powt run.
```
Comment (lgirdwood) [inline src/drivers/intel/hda/hda-dma.c:316]: typo "powt" ?

**Comment by plbossart** [comment]
that's the third PR on DMA L1 exit, with https://github.com/thesofproject/sof/pull/3135  merged 16 days ago.

Can we please have more background and commit messages that are clearer on what the problems are? This is just confusing, sorry.

File: src/drivers/intel/hda/hda-dma.c (hunk 1 above)
Comment (keyonjie) [reply src/drivers/intel/hda/hda-dma.c:316]: Thanks.

**Comment by keyonjie** [comment]
> that's the third PR on DMA L1 exit, with #3135 merged 16 days ago.

#3135 was merged to the release branch for the TGL-007 release. #3130 was the corresponding one for the master branch. Here #3223 is the replacement of #3130, which is for better solution that with extending the s0ix residency but preserving
the optimization from @tlauda of the low MCPS usage of the system DMA L1 exit. 

> 
> Can we please have more background and commit messages that are clearer on what the problems are? This is just confusing, sorry.

I thought I had added enough information in the long commit messages, looks still not enough?

File: src/drivers/intel/hda/hda-dma.c [hunk 2]
```diff
@@ -309,9 +311,14 @@ static void hda_dma_post_copy(struct dma_chan_data *chan, int bytes)
...
+		 */
 		if (!hda_chan->irq_disabled)
 			pm_runtime_put(PM_RUNTIME_HOST_DMA_L1, 0);
```
Comment (ranj063) [inline src/drivers/intel/hda/hda-dma.c:319]: @keyonjie we do a pm_runtime_get() on copy without the check for irq_disabled. So I wondering if we still need the check to do the pm_runtime_put() here ?

**Comment by plbossart** [comment]
> > that's the third PR on DMA L1 exit, with #3135 merged 16 days ago.
> 
> #3135 was merged to the release branch for the TGL-007 release. #3130 was the corresponding one for the master branch. Here #3223 is the replacement of #3130, which is for better solution that with extending the s0ix residency but preserving
> the optimization from @tlauda of the low MCPS usage of the system DMA L1 exit.
> 
> > Can we please have more background and commit messages that are clearer on what the problems are? This is just confusing, sorry.
> 
> I thought I had added enough information in the long commit messages, looks still not enough?

if you use the same commit message for a different solution it's very confusing.

Also doesn't this mean that the TGL-007 release includes a patch that's wrong and does not preserve the optimization?

File: src/drivers/intel/hda/hda-dma.c (hunk 2 above)
Comment (keyonjie) [reply src/drivers/intel/hda/hda-dma.c:319]: @ranj063 yes, we still need the check and do _put() in different ways for DMA IRQ driver and Timer driver scenarios, that is same with what @tlauda implemented. The only change is postpone the _get() for timer driver case.

**Comment by keyonjie** [comment]
> > > that's the third PR on DMA L1 exit, with #3135 merged 16 days ago.
> > 
> > 
> > #3135 was merged to the release branch for the TGL-007 release. #3130 was the corresponding one for the master branch. Here #3223 is the replacement of #3130, which is for better solution that with extending the s0ix residency but preserving
> > the optimization from @tlauda of the low MCPS usage of the system DMA L1 exit.
> > > Can we please have more background and commit messages that are clearer on what the problems are? This is just confusing, sorry.
> > 
> > 
> > I thought I had added enough information in the long commit messages, looks still not enough?
> 
> if you use the same commit message for a different solution it's very confusing.

No, it is new refined one here for this new solution.

> 
> Also doesn't this mean that the TGL-007 release includes a patch that's wrong and does not preserve the optimization?

That's partially true, the fact that extending the s0ix residency(functional gap) is more urgent than the MCPS performance lost in simultaneous pipelines scenarios, is why we delivered the release in a not perfect way.

**Comment by mmaka1** [comment]
SOFCI TEST

**Pull request merged**